File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque
def bfs(graph, i,j):
  dx = (-1, +1, 0, 0)
  dy = (0, 0, -1, +1)
  queue = deque()
  queue.append((i,j))
  graph[i][j] = 0
  count = 1
  while queue:
    x,y = map(int,queue.popleft())
    for k in range(4):
      nx = x + dx[k]
      ny = y + dy[k]
      if not(nx<n and ny<n and nx>-1 and ny>-1):
        continue
      if graph[nx][ny]==1:
        queue.append((nx, ny))
        graph[nx][ny] = 0
        count+=1
  return count

n = int(input())
graph = [[] for _ in range(n)] 
for i in range(n):
  for j in input():
    graph[i].append(int(j))

count = []
for i in range(n):
  for j in range(n):
    if graph[i][j]==1:
      count.append(bfs(graph,i,j))

count.sort()
print(len(count))
print("\n".join(str(i) for i in count))

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.io.*;
public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int[][] d = {{1,0},{-1,0},{0,1},{0,-1}};
    static int map[][];
    public static void main(String[] args) throws IOException {
        ArrayList<Integer> unit = new ArrayList<>();
        int n = Integer.parseInt(br.readLine());
        map = new int[n][n];
        for(int i = 0; i < n; i++){
            String s = br.readLine();
            for(int j = 0; j < n; j++){
                map[i][j] = s.charAt(j) - '0';
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if(map[i][j] == 1){
                    unit.add(bfs(i,j));
                }
            }
        }
        Collections.sort(unit);
        bw.write(unit.size()+"\n");
        for(int i : unit){
            bw.write(i+"\n");
        }
        bw.flush();
    }

    static int bfs(int startX, int startY){
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{startX, startY});
        map[startX][startY] = 2; //방문
        int cnt = 1;

        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            for(int i = 0; i < 4; i++){
                int nx = cur[0] + d[i][0];
                int ny = cur[1] + d[i][1];
                if(nx < 0 || nx >= map.length || ny < 0 || ny >= map[0].length)
                    continue;
                if(map[nx][ny] != 1)
                    continue;
                map[nx][ny] = 2;
                queue.add(new int[]{nx, ny});
                cnt++;
            }

        }
        return cnt;
    }
}
